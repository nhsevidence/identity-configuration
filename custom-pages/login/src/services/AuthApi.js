import Auth0 from 'auth0-js'
// import CordovaAuth0Plugin from 'auth0-js/dist/cordova-auth0-plugin.min'
import pathOr from 'ramda/src/pathOr'
import { auth as authOpts } from './constants'

export default class AuthApi {
  static instance = null
  constructor() {
    if (!window.config) {
      window.config = {}
    }
    window.config.extraParams = window.config.extraParams || {}
    this.opts = {
      domain: authOpts.domain,
      clientID: authOpts.clientID,
      // plugins: [new CordovaAuth0Plugin()],
      leeway: 1,
      popup: false,
      responseType: authOpts.responseType || 'code',
      scope: authOpts.scope,
      redirect: true
      // overrides: {
      //   // eslint-disable-next-line
      //   __tenant: config.auth0Tenant,
      //   // eslint-disable-next-line
      //   __token_issuer: config.authorizationServer.issuer
      // }
    }
    const params = Object.assign(this.opts, window.config.internalOptions)
    this.instance = new Auth0(params)
  }

  createAuth0Namespace = (promiseResolver) => {
    window.Auth0 = {
      setClient: (res) => {
        window.Auth0 = {
          ...window.Auth0,
          internalSettings: res,
          strategies: this.getStategies(res.strategies)
        }
        promiseResolver()
      }
    }
  }

  fetchClientSettings = () =>
    new Promise((resolver) => {
      if (window.config) {
        let source
        if (document.location.host.indexOf('localhost') !== -1) {
          source = `https://cdn.eu.auth0.com/client/ETzPLUtLTkCs8tHDjBfxNJKnnUzQGlmf.js?t${+new Date()}`
        } else {
          source = `${window.config.clientConfigurationBaseUrl}client/${
            authOpts.clientID
          }.js?t${+new Date()}`
        }
        this.createAuth0Namespace(resolver)
        const scriptTag = document.createElement('script')
        scriptTag.src = source
        document.body.appendChild(scriptTag)
      }
    })

  getStategies = strategies =>
    strategies.reduce((acc, curr) => {
      acc[curr.name] = {
        ...curr,
        domainString: curr.connections.length
          ? curr.connections[0].domain
          : null,
        connectionName: curr.connections.length
          ? curr.connections[0].name
          : null
      }
      return acc
    }, {})

  login(connection, email, password, errorCallback) {
    const redirectUri = pathOr(
      null,
      ['internalSettings', 'callback'],
      window.Auth0
    )
    let options
    let method
    if (connection === authOpts.connection) {
      options = {
        realm: connection,
        responseType: authOpts.responseType,
        email,
        password
      }
      method = 'login'
    } else {
      options = {
        connection,
        responseType: authOpts.responseType,
        email,
        sso: true,
        login_hint: email,
        response_mode: 'form_post'
      }
      if (redirectUri) {
        options.redirect_uri = redirectUri
      }
      method = 'authorize'
    }
    this.instance[method](options, (err) => {
      if (err) {
        if (errorCallback) {
          setTimeout(() =>
            errorCallback(method === 'login'
              ? 'Invalid email or password'
              : 'Something has gone wrong'))
        }
        throw new Error(err)
      }
    })
  }

  forgotPassword(email, errorCallback) {
    this.instance.changePassword(
      {
        connection: authOpts.connection,
        responseType: authOpts.responseType,
        email
      },
      (err) => {
        if (err) {
          if (errorCallback) {
            setTimeout(
              () =>
                errorCallback('There has been an issue, try a different email'),
              5
            )
          }
          return false
        }
        document.location.hash = '#/forgotsuccess'
        return true
      }
    )
  }

  resetPassword = (password, errorCallback) => {
    console.debug(window.rpConfig)
    if (window.rpConfig) {
      const data = {
        connection: authOpts.connection,
        responseType: authOpts.responseType,
        email: window.rpConfig.email,
        _csrf: window.rpConfig.csrf_token,
        ticket: window.rpConfig.ticket,
        newPassword: password,
        confirmNewPassword: password
      }

      fetch('/lo/reset', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      })
        .then((res) => {
          if (res.status === 200) {
            document.location.hash = '#/resetsuccess'
          } else if (errorCallback) {
            setTimeout(() => errorCallback('There has been an issue'), 5)
          }
        })
        .catch((err) => {
          if (errorCallback) {
            setTimeout(() => errorCallback('There has been an issue'), 5)
          }
          throw err
        })
    }
  }

  register(email, password, name, surname, allowContactMe, errorCallback) {
    return this.instance.signup(
      {
        connection: authOpts.connection,
        responseType: authOpts.responseType,
        email,
        password,
        user_metadata: {
          name,
          surname,
          allowContactMe
        }
      },
      (err) => {
        if (err) {
          if (errorCallback) {
            errorCallback()
          }
          return false
        }
        document.location.hash = '#/regsuccess'
        return true
      }
    )
  }
}
