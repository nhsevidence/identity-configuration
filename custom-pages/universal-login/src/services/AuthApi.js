/* eslint-disable no-underscore-dangle */
/* eslint-disable no-undef */
import Auth0 from 'auth0-js'
import qs from 'qs'
import { auth as authOpts, urls } from './constants'
import { ensureTrailingSlash } from '../helpers'

const __DEV__ = global.__DEV__ || false
export default class AuthApi {
  static instance = null
  constructor() {
    if (!window.config) {
      window.config = {authorizationServer: {}}
    }
    window.config.extraParams = window.config.extraParams || { redirectURI: undefined }
    this.clientID = window.config.clientID || authOpts.clientID
    this.opts = {
      ...window.config.extraParams,
      domain: authOpts.domain,
      clientID: this.clientID ,
      leeway: 1,
      popup: false,
      responseType: 'code',
      scope: authOpts.scope,
      redirect: true,
      configurationBaseUrl: config.clientConfigurationBaseUrl,
      overrides: {
        // eslint-disable-next-line
        __tenant: config.auth0Tenant,
        // eslint-disable-next-line
        __token_issuer: config.authorizationServer.issuer
      }
    }
    this.params = Object.assign(this.opts, window.config.internalOptions)
    this.instance = new Auth0.WebAuth(this.params)
  }

  createAuth0Namespace = (promiseResolver) => {
    window.Auth0 = {
      setClient: (res) => {
        window.Auth0 = {
          ...window.Auth0,
          internalSettings: res,
          strategies: this.getStategies(res.strategies)
        }
        promiseResolver()
      }
    }
  }

  fetchClientSettings = () =>
    new Promise((resolver) => {
      if (window.config) {
        const cdnBaseUrl = __DEV__
          ? authOpts.auth0CDN
          : window.config.clientConfigurationBaseUrl
        const source = `${ensureTrailingSlash(cdnBaseUrl)}client/${
          this.clientID
        }.js?t${+new Date()}`
        this.createAuth0Namespace(resolver)
        const scriptTag = document.createElement('script')
        scriptTag.src = source
        document.body.appendChild(scriptTag)
      }
    })

  getStategies = strategies =>
    strategies.reduce((acc, curr) => {
      acc[curr.name] = {
        ...curr,
        domainString: curr.connections.length
          ? curr.connections[0].domain
          : null,
        connectionName: curr.connections.length
          ? curr.connections[0].name
          : null
      }
      return acc
    }, {})

  login(connection, username, password, errorCallback, resumeAuthState) {
    console.log('Start login method');
    console.log(this.getCookie('_tempCid'));
    try {
      const redirectUri = window.config.extraParams.redirectURI
      const redirectUriWithGAId = `${redirectUri}?tempCid=2002` 
      let options
      let method
      if (connection === authOpts.connection) {
        options = {
          ...this.params,
          realm: connection,
          username,
          password,
          tempCid: '98765'
        }
        method = 'login'
      } else {
        options = {
          ...this.params,
          connection,
          username,
          sso: true,
          login_hint: username,
          response_mode: 'form_post',
          tempCid: '123456789S'
        }
        method = 'authorize'
      }
      if (redirectUri) {
        console.log('redirectUri set');
        options.redirect_uri = `${redirectUri}?tempCid=1001` 
        console.log(`Redirect url = ${options.redirect_uri}`);
      } else {
        options.redirect_uri = `https://local-identity.nice.org.uk/?tempCid=333` 
      }

      if (!resumeAuthState) {
        console.log('not resumeAuthState');
        console.log(`method ${method}`);
        console.log(`options redirect url ${JSON.stringify(options.redirect_uri)}`);

        this.instance[method](options, (err, result) => {
          if (result)
          {
            console.log(`result = ${JSON.stringify(result)}`);
          }
          if (err) {
            console.log(`error occured ${err}`);
            if (errorCallback) {
              setTimeout(() => errorCallback(err))
            }
            console.log(JSON.stringify(err))
          }
        })
      } else {
        console.log('resumeAuthState');
        const GETOptions = qs.stringify(
          { ...options, state: resumeAuthState, tempCid: 123 },
          { addQueryPrefix: true }
        )
        fetch(`/continue${GETOptions}`, {
          method: 'GET',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json'
          }
        })
          .then((res) => {
            if (res.status === 200) {
              console.log(`redirectUri = ${redirectUriWithGAId}`);
              document.location = redirectUriWithGAId
            } else if (errorCallback) {
              setTimeout(() => errorCallback(res))
            }
          })
          .catch((err) => {
            console.log('Error occured');
            if (errorCallback) {
              setTimeout(() => errorCallback(err))
            }
          })
      }
    } catch (err) {
      console.log(JSON.stringify(err))
    }
  }

  submitWSForm = (responseForm) => {
    const div = document.createElement('div')
    div.innerHTML = responseForm
    const formElement = document.body.appendChild(div).children[0]
    formElement.submit()
  }

  forgotPassword(email, errorCallback, history) {
    const options = {
      ...this.params,
      connection: authOpts.connection,
      responseType: authOpts.responseType,
      email
    }
    this.instance.changePassword(options, (err) => {
      if (err) {
        if (errorCallback) {
          setTimeout(() => errorCallback(err), 5)
        }
        return false
      }
      history.push('/forgotsuccess');
      return true
    })    
  }

  resetPassword = (password, errorCallback, history) => {
    const callback = (res) => {
      if (res.status === 200) {
        history.push('/resetsuccess');
      } else if (errorCallback) {
        setTimeout(() => errorCallback('There has been an issue'))
      }
    }
    const catchCallback = (err) => {
      if (errorCallback) {
        setTimeout(() => errorCallback('There has been an issue'))
      }
      console.log(JSON.stringify(err))
    }

    if (window.rpConfig) {
      const data = {
        connection: authOpts.connection,
        responseType: authOpts.responseType,
        email: window.rpConfig.email,
        _csrf: window.rpConfig.csrf_token,
        ticket: window.rpConfig.ticket,
        newPassword: password,
        confirmNewPassword: password
      }
      
      fetch('/lo/reset', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      })
        .then(callback)
        .catch(catchCallback)      
    }
  }

  register(email, password, name, surname, acceptedTerms, allowContactMe, errorCallback, history) {
    const options = {
      ...this.params,
      connection: authOpts.connection,
      responseType: authOpts.responseType,
      email,
      password,
      user_metadata: {
        name,
        surname,
        acceptedTerms,
        allowContactMe
      }
    }
    console.log(options);
    this.instance.signup(options, (err) => {
      if (err) {
        console.log(err);
        if (errorCallback) {
          setTimeout(() => errorCallback(err))
        }
        return false
      }
      history.push('/regsuccess');
      return true
    })
  }

  resendVerificationEmail = (userId, callerCallback) => {
    console.log('resending email to: ', userId)
    const callback = (res) => {
      console.log(JSON.stringify(res));
      console.log(res.status);
      if (res.status === 200) {
        setTimeout(() => callerCallback(null)) // this will reset the error
      } else {
        setTimeout(() =>
          callerCallback('something has gone wrong when sending the email'))
      }
    }
    const catchCallback = (err) => {
      setTimeout(() =>
        callerCallback('something has gone quite wrong when sending the email'))
      console.log(JSON.stringify(err))
    }

    const data = {
      user_id: userId
    }
    console.log('about to resend verification email');
    console.log(`url: ${urls.resendVerificationEmail} body: ${JSON.stringify(data)}`);      
    fetch(urls.resendVerificationEmail, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data)
    })
      .then(callback)
      .catch(catchCallback)
  }

  getCookie = (name) => {
    console.log(`Start get cookie - ${name}`);
    const regx = new RegExp(name + '=([^;]+)')
    const match = document.cookie.match(regx);
    console.log(`cookie match = ${match}`);
    if (match) {
      console.log('Found Cookie');
      console.log(match[1]);
    } else {
      console.log(`something has gone wrong when getting the cookie - ${regx}`);
    }
    console.log(`End get cookie - ${name}`);
  }
}