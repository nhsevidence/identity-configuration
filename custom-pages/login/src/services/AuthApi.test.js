import AuthApi from './AuthApi'


describe('AppContainer components', () => {
  it('should create an instance of AuthApi correctly', () => {
    expect(new AuthApi()).toMatchSnapshot()
  })
})
