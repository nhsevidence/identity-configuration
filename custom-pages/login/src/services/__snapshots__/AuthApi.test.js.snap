// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer components should create an instance of AuthApi correctly 1`] = `
AuthApi {
  "createAuth0Namespace": [Function],
  "fetchClientSettings": [Function],
  "forgotPasswordIE8": [Function],
  "getStategies": [Function],
  "instance": WebAuth {
    "_universalLogin": HostedPages {
      "baseOptions": Object {
        "_sendTelemetry": true,
        "_timesToRetryFailedRequests": 0,
        "clientID": "#{AUTH0_APP_CLIENT_ID}",
        "domain": "#{AUTH0_DOMAIN}",
        "jwksURI": undefined,
        "leeway": 1,
        "plugins": PluginHandler {
          "plugins": Array [],
        },
        "popup": false,
        "popupOrigin": undefined,
        "redirect": true,
        "responseType": "code",
        "rootUrl": "https://#{AUTH0_DOMAIN}",
        "scope": "openid profile email update:users",
        "tenant": "#{AUTH0_DOMAIN}",
        "token_issuer": "https://#{AUTH0_DOMAIN}/",
      },
      "client": [Circular],
      "request": RequestBuilder {
        "_sendTelemetry": true,
        "_telemetryInfo": null,
        "_timesToRetryFailedRequests": 0,
        "headers": Object {},
      },
      "warn": Warn {
        "disableWarnings": false,
      },
    },
    "baseOptions": Object {
      "_sendTelemetry": true,
      "_timesToRetryFailedRequests": 0,
      "clientID": "#{AUTH0_APP_CLIENT_ID}",
      "domain": "#{AUTH0_DOMAIN}",
      "jwksURI": undefined,
      "leeway": 1,
      "plugins": PluginHandler {
        "plugins": Array [],
      },
      "popup": false,
      "popupOrigin": undefined,
      "redirect": true,
      "responseType": "code",
      "rootUrl": "https://#{AUTH0_DOMAIN}",
      "scope": "openid profile email update:users",
      "tenant": "#{AUTH0_DOMAIN}",
      "token_issuer": "https://#{AUTH0_DOMAIN}/",
    },
    "client": Authentication {
      "baseOptions": Object {
        "_sendTelemetry": true,
        "_timesToRetryFailedRequests": 0,
        "clientID": "#{AUTH0_APP_CLIENT_ID}",
        "domain": "#{AUTH0_DOMAIN}",
        "jwksURI": undefined,
        "leeway": 1,
        "plugins": PluginHandler {
          "plugins": Array [],
        },
        "popup": false,
        "popupOrigin": undefined,
        "redirect": true,
        "responseType": "code",
        "rootUrl": "https://#{AUTH0_DOMAIN}",
        "scope": "openid profile email update:users",
        "tenant": "#{AUTH0_DOMAIN}",
        "token_issuer": "https://#{AUTH0_DOMAIN}/",
      },
      "dbConnection": DBConnection {
        "baseOptions": Object {
          "_sendTelemetry": true,
          "_timesToRetryFailedRequests": 0,
          "clientID": "#{AUTH0_APP_CLIENT_ID}",
          "domain": "#{AUTH0_DOMAIN}",
          "jwksURI": undefined,
          "leeway": 1,
          "plugins": PluginHandler {
            "plugins": Array [],
          },
          "popup": false,
          "popupOrigin": undefined,
          "redirect": true,
          "responseType": "code",
          "rootUrl": "https://#{AUTH0_DOMAIN}",
          "scope": "openid profile email update:users",
          "tenant": "#{AUTH0_DOMAIN}",
          "token_issuer": "https://#{AUTH0_DOMAIN}/",
        },
        "request": RequestBuilder {
          "_sendTelemetry": true,
          "_telemetryInfo": null,
          "_timesToRetryFailedRequests": 0,
          "headers": Object {},
        },
      },
      "passwordless": PasswordlessAuthentication {
        "baseOptions": Object {
          "_sendTelemetry": true,
          "_timesToRetryFailedRequests": 0,
          "clientID": "#{AUTH0_APP_CLIENT_ID}",
          "domain": "#{AUTH0_DOMAIN}",
          "jwksURI": undefined,
          "leeway": 1,
          "plugins": PluginHandler {
            "plugins": Array [],
          },
          "popup": false,
          "popupOrigin": undefined,
          "redirect": true,
          "responseType": "code",
          "rootUrl": "https://#{AUTH0_DOMAIN}",
          "scope": "openid profile email update:users",
          "tenant": "#{AUTH0_DOMAIN}",
          "token_issuer": "https://#{AUTH0_DOMAIN}/",
        },
        "request": RequestBuilder {
          "_sendTelemetry": true,
          "_telemetryInfo": null,
          "_timesToRetryFailedRequests": 0,
          "headers": Object {},
        },
      },
      "request": RequestBuilder {
        "_sendTelemetry": true,
        "_telemetryInfo": null,
        "_timesToRetryFailedRequests": 0,
        "headers": Object {},
      },
      "ssodataStorage": SSODataStorage {
        "storage": Storage {
          "handler": StorageHandler {
            "storage": CookieStorage {},
            "warn": Warn {
              "disableWarnings": undefined,
            },
          },
        },
      },
      "warn": Warn {
        "disableWarnings": false,
      },
    },
    "crossOriginAuthentication": CrossOriginAuthentication {
      "baseOptions": Object {
        "_sendTelemetry": true,
        "_timesToRetryFailedRequests": 0,
        "clientID": "#{AUTH0_APP_CLIENT_ID}",
        "domain": "#{AUTH0_DOMAIN}",
        "jwksURI": undefined,
        "leeway": 1,
        "plugins": PluginHandler {
          "plugins": Array [],
        },
        "popup": false,
        "popupOrigin": undefined,
        "redirect": true,
        "responseType": "code",
        "rootUrl": "https://#{AUTH0_DOMAIN}",
        "scope": "openid profile email update:users",
        "tenant": "#{AUTH0_DOMAIN}",
        "token_issuer": "https://#{AUTH0_DOMAIN}/",
      },
      "request": RequestBuilder {
        "_sendTelemetry": true,
        "_telemetryInfo": null,
        "_timesToRetryFailedRequests": 0,
        "headers": Object {},
      },
      "storage": Storage {
        "handler": StorageHandler {
          "storage": CookieStorage {},
          "warn": Warn {
            "disableWarnings": undefined,
          },
        },
      },
      "webAuth": [Circular],
      "webMessageHandler": WebMessageHandler {
        "warn": Warn {
          "disableWarnings": undefined,
        },
        "webAuth": [Circular],
      },
    },
    "popup": Popup {
      "baseOptions": Object {
        "_sendTelemetry": true,
        "_timesToRetryFailedRequests": 0,
        "clientID": "#{AUTH0_APP_CLIENT_ID}",
        "domain": "#{AUTH0_DOMAIN}",
        "jwksURI": undefined,
        "leeway": 1,
        "plugins": PluginHandler {
          "plugins": Array [],
        },
        "popup": false,
        "popupOrigin": undefined,
        "redirect": true,
        "responseType": "code",
        "rootUrl": "https://#{AUTH0_DOMAIN}",
        "scope": "openid profile email update:users",
        "tenant": "#{AUTH0_DOMAIN}",
        "token_issuer": "https://#{AUTH0_DOMAIN}/",
      },
      "client": Authentication {
        "baseOptions": Object {
          "_sendTelemetry": true,
          "_timesToRetryFailedRequests": 0,
          "clientID": "#{AUTH0_APP_CLIENT_ID}",
          "domain": "#{AUTH0_DOMAIN}",
          "jwksURI": undefined,
          "leeway": 1,
          "plugins": PluginHandler {
            "plugins": Array [],
          },
          "popup": false,
          "popupOrigin": undefined,
          "redirect": true,
          "responseType": "code",
          "rootUrl": "https://#{AUTH0_DOMAIN}",
          "scope": "openid profile email update:users",
          "tenant": "#{AUTH0_DOMAIN}",
          "token_issuer": "https://#{AUTH0_DOMAIN}/",
        },
        "dbConnection": DBConnection {
          "baseOptions": Object {
            "_sendTelemetry": true,
            "_timesToRetryFailedRequests": 0,
            "clientID": "#{AUTH0_APP_CLIENT_ID}",
            "domain": "#{AUTH0_DOMAIN}",
            "jwksURI": undefined,
            "leeway": 1,
            "plugins": PluginHandler {
              "plugins": Array [],
            },
            "popup": false,
            "popupOrigin": undefined,
            "redirect": true,
            "responseType": "code",
            "rootUrl": "https://#{AUTH0_DOMAIN}",
            "scope": "openid profile email update:users",
            "tenant": "#{AUTH0_DOMAIN}",
            "token_issuer": "https://#{AUTH0_DOMAIN}/",
          },
          "request": RequestBuilder {
            "_sendTelemetry": true,
            "_telemetryInfo": null,
            "_timesToRetryFailedRequests": 0,
            "headers": Object {},
          },
        },
        "passwordless": PasswordlessAuthentication {
          "baseOptions": Object {
            "_sendTelemetry": true,
            "_timesToRetryFailedRequests": 0,
            "clientID": "#{AUTH0_APP_CLIENT_ID}",
            "domain": "#{AUTH0_DOMAIN}",
            "jwksURI": undefined,
            "leeway": 1,
            "plugins": PluginHandler {
              "plugins": Array [],
            },
            "popup": false,
            "popupOrigin": undefined,
            "redirect": true,
            "responseType": "code",
            "rootUrl": "https://#{AUTH0_DOMAIN}",
            "scope": "openid profile email update:users",
            "tenant": "#{AUTH0_DOMAIN}",
            "token_issuer": "https://#{AUTH0_DOMAIN}/",
          },
          "request": RequestBuilder {
            "_sendTelemetry": true,
            "_telemetryInfo": null,
            "_timesToRetryFailedRequests": 0,
            "headers": Object {},
          },
        },
        "request": RequestBuilder {
          "_sendTelemetry": true,
          "_telemetryInfo": null,
          "_timesToRetryFailedRequests": 0,
          "headers": Object {},
        },
        "ssodataStorage": SSODataStorage {
          "storage": Storage {
            "handler": StorageHandler {
              "storage": CookieStorage {},
              "warn": Warn {
                "disableWarnings": undefined,
              },
            },
          },
        },
        "warn": Warn {
          "disableWarnings": false,
        },
      },
      "crossOriginAuthentication": CrossOriginAuthentication {
        "baseOptions": Object {
          "_sendTelemetry": true,
          "_timesToRetryFailedRequests": 0,
          "clientID": "#{AUTH0_APP_CLIENT_ID}",
          "domain": "#{AUTH0_DOMAIN}",
          "jwksURI": undefined,
          "leeway": 1,
          "plugins": PluginHandler {
            "plugins": Array [],
          },
          "popup": false,
          "popupOrigin": undefined,
          "redirect": true,
          "responseType": "code",
          "rootUrl": "https://#{AUTH0_DOMAIN}",
          "scope": "openid profile email update:users",
          "tenant": "#{AUTH0_DOMAIN}",
          "token_issuer": "https://#{AUTH0_DOMAIN}/",
        },
        "request": RequestBuilder {
          "_sendTelemetry": true,
          "_telemetryInfo": null,
          "_timesToRetryFailedRequests": 0,
          "headers": Object {},
        },
        "storage": Storage {
          "handler": StorageHandler {
            "storage": CookieStorage {},
            "warn": Warn {
              "disableWarnings": undefined,
            },
          },
        },
        "webAuth": [Circular],
        "webMessageHandler": WebMessageHandler {
          "warn": Warn {
            "disableWarnings": undefined,
          },
          "webAuth": [Circular],
        },
      },
      "transactionManager": TransactionManager {
        "keyLength": 32,
        "namespace": "com.auth0.auth.",
        "options": Object {
          "_sendTelemetry": true,
          "_timesToRetryFailedRequests": 0,
          "clientID": "#{AUTH0_APP_CLIENT_ID}",
          "domain": "#{AUTH0_DOMAIN}",
          "jwksURI": undefined,
          "leeway": 1,
          "plugins": PluginHandler {
            "plugins": Array [],
          },
          "popup": false,
          "popupOrigin": undefined,
          "redirect": true,
          "responseType": "code",
          "rootUrl": "https://#{AUTH0_DOMAIN}",
          "scope": "openid profile email update:users",
          "tenant": "#{AUTH0_DOMAIN}",
          "token_issuer": "https://#{AUTH0_DOMAIN}/",
        },
        "storage": Storage {
          "handler": StorageHandler {
            "storage": CookieStorage {},
            "warn": Warn {
              "disableWarnings": undefined,
            },
          },
        },
      },
      "warn": Warn {
        "disableWarnings": false,
      },
      "webAuth": [Circular],
    },
    "redirect": Redirect {
      "baseOptions": Object {
        "_sendTelemetry": true,
        "_timesToRetryFailedRequests": 0,
        "clientID": "#{AUTH0_APP_CLIENT_ID}",
        "domain": "#{AUTH0_DOMAIN}",
        "jwksURI": undefined,
        "leeway": 1,
        "plugins": PluginHandler {
          "plugins": Array [],
        },
        "popup": false,
        "popupOrigin": undefined,
        "redirect": true,
        "responseType": "code",
        "rootUrl": "https://#{AUTH0_DOMAIN}",
        "scope": "openid profile email update:users",
        "tenant": "#{AUTH0_DOMAIN}",
        "token_issuer": "https://#{AUTH0_DOMAIN}/",
      },
      "crossOriginAuthentication": CrossOriginAuthentication {
        "baseOptions": Object {
          "_sendTelemetry": true,
          "_timesToRetryFailedRequests": 0,
          "clientID": "#{AUTH0_APP_CLIENT_ID}",
          "domain": "#{AUTH0_DOMAIN}",
          "jwksURI": undefined,
          "leeway": 1,
          "plugins": PluginHandler {
            "plugins": Array [],
          },
          "popup": false,
          "popupOrigin": undefined,
          "redirect": true,
          "responseType": "code",
          "rootUrl": "https://#{AUTH0_DOMAIN}",
          "scope": "openid profile email update:users",
          "tenant": "#{AUTH0_DOMAIN}",
          "token_issuer": "https://#{AUTH0_DOMAIN}/",
        },
        "request": RequestBuilder {
          "_sendTelemetry": true,
          "_telemetryInfo": null,
          "_timesToRetryFailedRequests": 0,
          "headers": Object {},
        },
        "storage": Storage {
          "handler": StorageHandler {
            "storage": CookieStorage {},
            "warn": Warn {
              "disableWarnings": undefined,
            },
          },
        },
        "webAuth": [Circular],
        "webMessageHandler": WebMessageHandler {
          "warn": Warn {
            "disableWarnings": undefined,
          },
          "webAuth": [Circular],
        },
      },
      "warn": Warn {
        "disableWarnings": false,
      },
      "webAuth": [Circular],
    },
    "ssodataStorage": SSODataStorage {
      "storage": Storage {
        "handler": StorageHandler {
          "storage": CookieStorage {},
          "warn": Warn {
            "disableWarnings": undefined,
          },
        },
      },
    },
    "transactionManager": TransactionManager {
      "keyLength": 32,
      "namespace": "com.auth0.auth.",
      "options": Object {
        "_sendTelemetry": true,
        "_timesToRetryFailedRequests": 0,
        "clientID": "#{AUTH0_APP_CLIENT_ID}",
        "domain": "#{AUTH0_DOMAIN}",
        "jwksURI": undefined,
        "leeway": 1,
        "plugins": PluginHandler {
          "plugins": Array [],
        },
        "popup": false,
        "popupOrigin": undefined,
        "redirect": true,
        "responseType": "code",
        "rootUrl": "https://#{AUTH0_DOMAIN}",
        "scope": "openid profile email update:users",
        "tenant": "#{AUTH0_DOMAIN}",
        "token_issuer": "https://#{AUTH0_DOMAIN}/",
      },
      "storage": Storage {
        "handler": StorageHandler {
          "storage": CookieStorage {},
          "warn": Warn {
            "disableWarnings": undefined,
          },
        },
      },
    },
    "webMessageHandler": WebMessageHandler {
      "warn": Warn {
        "disableWarnings": undefined,
      },
      "webAuth": [Circular],
    },
  },
  "loginIE8": [Function],
  "opts": Object {
    "_sendTelemetry": true,
    "_timesToRetryFailedRequests": 0,
    "clientID": "#{AUTH0_APP_CLIENT_ID}",
    "domain": "#{AUTH0_DOMAIN}",
    "jwksURI": undefined,
    "leeway": 1,
    "plugins": PluginHandler {
      "plugins": Array [],
    },
    "popup": false,
    "popupOrigin": undefined,
    "redirect": true,
    "responseType": "code",
    "rootUrl": "https://#{AUTH0_DOMAIN}",
    "scope": "openid profile email update:users",
    "tenant": "#{AUTH0_DOMAIN}",
    "token_issuer": "https://#{AUTH0_DOMAIN}/",
  },
  "params": Object {
    "_sendTelemetry": true,
    "_timesToRetryFailedRequests": 0,
    "clientID": "#{AUTH0_APP_CLIENT_ID}",
    "domain": "#{AUTH0_DOMAIN}",
    "jwksURI": undefined,
    "leeway": 1,
    "plugins": PluginHandler {
      "plugins": Array [],
    },
    "popup": false,
    "popupOrigin": undefined,
    "redirect": true,
    "responseType": "code",
    "rootUrl": "https://#{AUTH0_DOMAIN}",
    "scope": "openid profile email update:users",
    "tenant": "#{AUTH0_DOMAIN}",
    "token_issuer": "https://#{AUTH0_DOMAIN}/",
  },
  "registerIE8": [Function],
  "resendActivationEmail": [Function],
  "resetPassword": [Function],
  "submitWSForm": [Function],
}
`;
